pkg/checkpoint/protocol.go

```diff
@@ -125,277 +68,194 @@ func NewProtocol(
+			"memberships", len(params.StateSnapshot.EpochData.EpochConfig.Memberships),
+		)
+
+		// Apply pending Checkpoint messages
+		for s, msg := range params.PendingMessages {
+			if err := applyCheckpointReceived(s, msg.Epoch, msg.Sn, msg.SnapshotHash, msg.Signature); err != nil {
+				params.Log(logging.LevelWarn, "Error applying pending Checkpoint message", "error", err, "msg", msg)
+			}
 
-func (p *Protocol) applyMessage(msg *checkpointpb.Checkpoint, source t.NodeID) *events.EventList {
-	eventsOut := events.EmptyList()
+		}
+		params.PendingMessages = nil
 
-	// check if source is part of the membership
-	if !sliceutil.Contains(p.membership, source) {
-		p.Logger.Log(logging.LevelWarn, "sender %s is not a member.\n", source)
-		return events.EmptyList()
-	}
-	// Notify the protocol about the progress of the source node.
-	// If no progress is made for a configured number of epochs,
-	// the node is considered to be a straggler and is sent a stable checkpoint to catch up.
-	eventsOut.PushBack(protobufs.EpochProgressEvent(p.moduleConfig.Ord, source, tt.EpochNr(msg.Epoch)))
-
-	// If checkpoint is already stable, ignore message.
-	if p.stable() {
-		return eventsOut
-	}
+		return nil
+	})
 
-	// Check snapshot hash
-	if p.stateSnapshotHash == nil {
-		// The message is received too early, put it aside
-		p.pendingMessages[source] = msg
-		return eventsOut
-	} else if !bytes.Equal(p.stateSnapshotHash, msg.SnapshotHash) {
-		// Snapshot hash mismatch
-		p.Log(logging.LevelWarn, "Ignoring Checkpoint message. Mismatching state snapshot hash.", "source", source)
-		return eventsOut
-	}
+	cryptopbdsl.UponSigsVerified(m, func(nodeIds []t.NodeID, errors []error, allOk bool, _ *t.EmptyContext) error {
 
-	// TODO: Only accept messages from nodes in membership.
-	//       This might be more tricky than it seems, especially when the membership is not yet initialized.
+		// A checkpoint only has one signature and thus each slice of the result only contains one element.
+		sourceNode := nodeIds[0]
+		err := errors[0]
 
-	// Ignore duplicate messages.
-	if _, ok := p.signatures[source]; ok {
-		return eventsOut
-	}
-	p.signatures[source] = msg.Signature
-
-	// Verify signature of the sender.
-	sigData := serializeCheckpointForSig(p.epoch, p.seqNr, p.stateSnapshotHash)
-	eventsOut.PushBack(cryptopbevents.VerifySigs(
-		p.moduleConfig.Crypto,
-		[]*cryptopbtypes.SignedData{sigData},
-		[][]byte{msg.Signature},
-		protobufs.SigVerOrigin(p.moduleConfig.Self),
-		[]t.NodeID{source},
-	).Pb())
-
-	return eventsOut
-}
+		if !allOk {
+			params.Log(logging.LevelWarn, "Ignoring Checkpoint message. Invalid signature.", "source", sourceNode, "error", err)
+			params.Signatures[sourceNode] = nil
+			return nil
+		}
 
-func (p *Protocol) applyNodeSigsVerified(result *cryptopb.SigsVerified) (*events.EventList, error) {
+		// Note the reception of a valid Checkpoint message from node `source`.
+		params.Confirmations[sourceNode] = struct{}{}
```

Comment:
We cannot really leave `params.Signatures` only if we want to be able to ignore duplicates and keep clarity. The context does however allows extra clarity in that signatures can be populated only once the signature is verified (contrary to the current case). I created `params.SigReceived` and removed `params.Confirmations` which seems more natural.

Reply:
Yes, good point. I still think that the extra data structure is technically not required, because it only removes duplicates that arrive while the first received signature is being verified. So, in the worst case, a malicious node could flood us with `Checkpoint` messages and mount a DoS by making us uselessly verify a huge load of signatures. The extra data structure prevents that, which is nice, so let's keep it.